### PR TITLE
Add initial framework for integration testing.

### PR DESCRIPTION
**Pull request by kenzieschmoll**
This PR adds 

- integration_tests/test/infra/test_app_driver.dart : this code was copied from the existing file flutter_test_driver.dart, but paired down to use only what we need right now. This code will be used to spin up a test app that we will connect to DevTools for integration testing.
- test_driver/integration_test.dart : this is the test driver we will use for integration tests

### CODE REVIEW

**Comment by polina-c** [comment]
Why `test_driver/integration_test.dart` is not in test_infra?
Are you sure we want to have separate test_infra in integration_test? Would it be more convenient to have it shared between different types of testing? 

Couple options:
1. take test_infra out of test and integration_test and make it shared
2. Declare that test is lower level than integration_test and use test/test-infra for both

File: packages/devtools_app/integration_test/test_infra/test_app_driver.dart [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    String? event,
+    int? id,
+    Duration? timeout,
+    bool ignoreAppStopEvent = false,
+  }) async {
+    final Completer<Map<String, dynamic>> response =
+        Completer<Map<String, dynamic>>();
+    late StreamSubscription<String> sub;
+    sub = stdoutController.stream.listen((String line) async {
+      final dynamic json = _parseFlutterResponse(line);
```
Comment (bkonyi) [inline packages/devtools_app/integration_test/test_infra/test_app_driver.dart:143]: Consider pulling this closure into its own method.

Comment (bkonyi) [inline packages/devtools_app/integration_test/test_infra/test_app_driver.dart:143]: Is this type necessary? `_parseFlutterResponse` has a known, non-`dynamic` return type.

File: packages/devtools_app/integration_test/test_infra/test_app_driver.dart [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+      if (json == null) {
+        return;
+      } else if ((event != null && json['event'] == event) ||
```
Comment (bkonyi) [inline packages/devtools_app/integration_test/test_infra/test_app_driver.dart:146]: Nit: use constants for keys, here and elsewhere.

File: packages/devtools_app/integration_test/test_infra/test_app_driver.dart [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+          (id != null && json['id'] == id)) {
+        await sub.cancel();
+        response.complete(json);
+      } else if (!ignoreAppStopEvent && json['event'] == 'app.stop') {
```
Comment (bkonyi) [inline packages/devtools_app/integration_test/test_infra/test_app_driver.dart:150]: Do we have constants defined in `devtools_shared` for these daemon event names?

File: packages/devtools_app/integration_test/test_infra/test_app_driver.dart [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+  }
+
+  Future<T> _timeoutWithMessages<T>(
+    Future<T> Function() f, {
+    Duration? timeout,
+    String? message,
+  }) {
+    // Capture output to a buffer so if we don't get the response we want we can show
+    // the output that did arrive in the timeout error.
+    final StringBuffer messages = StringBuffer();
```
Comment (bkonyi) [inline packages/devtools_app/integration_test/test_infra/test_app_driver.dart:183]: Nit: consider removing LHS types for declarations where it's clear what the type should be.

File: packages/devtools_app/integration_test/test_infra/io_utils.dart [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import 'dart:convert';
+import 'dart:io';
+
+Stream<String> transformToLines(Stream<List<int>> byteStream) {
+  return byteStream
+      .transform<String>(utf8.decoder)
+      .transform<String>(const LineSplitter());
+}
+
+mixin IoMixin {
```
Comment (bkonyi) [inline packages/devtools_app/integration_test/test_infra/io_utils.dart:15]: Ubernit: `s/Io/IO/`

**Review by bkonyi — COMMENTED** [review]
DBR

**Comment by kenzieschmoll** [comment]
> Why `test_driver/integration_test.dart` is not in test_infra? Are you sure we want to have separate test_infra in integration_test? Would it be more convenient to have it shared between different types of testing?
> 
> Couple options:
> 
> 1. take test_infra out of test and integration_test and make it shared
> 2. Declare that test is lower level than integration_test and use test/test-infra for both

Official Flutter guidance recommends code structure:
```
lib/
  ...
integration_test/
test/
```

I did try to create a shared test folder, but libraries under test/ and integration_test/ could not import anything outside of their direct folders or lib/.

Unit tests and integration tests will have some different "test_infra" so it is okay to have two folders. Once we build out our integration tests more fully, it may make sense to put anything that could be shared between test/ and integration_test into packages/devtools_test so that both test/ and integration_test can share it. This can be done at a later time though when we have a better idea of what will need to be shared.

File: packages/devtools_app/integration_test/test_infra/test_app_driver.dart (hunk 1 above)
Comment (kenzieschmoll) [reply packages/devtools_app/integration_test/test_infra/test_app_driver.dart:143]: probably not. All this code was written in pre-null safe Dart land and copied over in this PR. updated.

File: packages/devtools_app/integration_test/test_infra/test_app_driver.dart (hunk 3 above)
Comment (kenzieschmoll) [reply packages/devtools_app/integration_test/test_infra/test_app_driver.dart:150]: created enum FlutterDaemonConstants for now with a TODO to consider placing these constants in devtools_shared or in flutter depending on where they would be used.

**Review by polina-c — APPROVED** [review]

**Comment by kenzieschmoll** [comment]
Work towards https://github.com/flutter/devtools/issues/4934

**Pull request merged**